README.md

```diff
@@ -1,61 +1,2354 @@
+directories and files. Every time the filesystem is modified in any way in the
+Dockerfile, a new layer is created. This usually occurs after each full
+directive.
+
+As well as inheriting from base images, it's also possible to start a completely
+new Docker image which does not include any layers. This can be achieved by
+using the special image name `scratch`. This tells Docker it should try to
+inherit from any other image, but start a brand new set of layers. In fact, this
+is how a large number of 'standard' base images start, such as the Docker
+versions of Linux distributions work, by starting from `scratch` and copying in
+all of the files that build up that version of Linux.
+
+#### 4.2 [`COPY`](https://docs.docker.com/engine/reference/builder/#copy) Directive
+
+So now we now how to use a base image to start with a filesystem in our own
+image, but how do we get new files into it?
+
+This is where the `COPY` directive comes in. This allows the copying of files
+from the host's local filesystem into the Docker image being built. It's fairly
+similar to the standard copying commands under most OS variants these days in
+that it takes the location of a source file or directory on the host and a
+destination where the contents of the source should be copied inside the image.
+`COPY` preserves all flags on the file when copied, so for example if you copy
+a shell script that has the execution flag set, this will be preserved when
+copied into the image.
+
+**NOTE:** The directory location of the Dockerfile being used to build the image
+    the image is known as the build context. Any files in the build context, or
+    any of it's sub-directory hierarchy is also part of the build context.
+    This can be changed by other commands, but this is the default and we'll be
+    using it for the moment.
+
+The following will recursively copy the contents of the  `directory-3` directory
+the current build context into the image at the location `/directory-3`:
+```
+COPY directory-3 /directory-3/
+```
+Note that you need to add the trailing `/` to the destination directory. If
+the destination directory doesn't already exist, it'll be dynamically created
+(which means you can specify a nested destination directory without having to
+run `mkdir` in the image).
+
+Wildcards also work, so that the following:
+```
+COPY * /all-the-files/
+```
+Will recursively copy *everything* (including the Dockerfile) in the current
+build context to the `/all-the-files/` directory in the built image.
+
+It should be noted that the source of files are *always* relative to the build
+context. Therefore files must be available to that context (you can't, for
+example, reference files that are in the directory above where the Dockerfile
+is, because it's not within the same context).
+
+`COPY` has a few switches to, that include abilities to set ownership and
+on files as they're copied. Refer to the Dockerfile documentation for more
```

Comment:
```suggestion
on files as they're copied. Refer to the [Dockerfile documentation](https://docs.docker.com/engine/reference/builder/#copy) for more
```

Reply:
Hmm, I'd actually put a hyperlink in the COPY at the top. Do you think it's worth duplicating (I don't want to end up putting hyperlinks everywhere, esp. if they get out of date).

Reply:
I tend to err on the side of adding a hyperlink for any text that is explicitly referencing something such as this. In this case, though I think the odds on that one getting broken would be slim, and trivial to resolve, if worried about duplicates you can always use reference links such that the link is only included once i.e.

```
Refer to the [Dockerfile documentation][docker-copy-ref] for more

[docker-copy-ref]: https://docs.docker.com/engine/reference/builder/#copy
```

Then you get into mixing inline and reference links so 🤷‍♂